### PR TITLE
Update submariner-charts version during the installers phase

### LINKED ISSUE #438: Releases should update the charts versions
See https://github.com/submariner-io/submariner-charts/pull/243, https://github.com/submariner-io/submariner-charts/pull/244, https://github.com/submariner-io/submariner-charts/pull/245 and https://github.com/submariner-io/submariner-charts/pull/246 for examples of what needs to be done.

### PR DESCRIPTION
**Pull request by tpantelis**
Adds a step to create a PR in **submariner-charts** to update the `CHARTS_VERSION` in the `Makefile`.

See https://github.com/submariner-io/submariner-charts/pull/379 which was generated locally via the changes in this PR,

Fixes https://github.com/submariner-io/releases/issues/438

### CODE REVIEW

**Review by dfarrell07 — APPROVED** [review]

**Review by skitt — CHANGES_REQUESTED** [review]
Shouldn’t this take place during the installers phase, so that the tagged Helm charts have the updated contents?

**Comment by tpantelis** [comment]
> Shouldn’t this take place during the installers phase, so that the tagged Helm charts have the updated contents?

I'm not sure what you mean by that. The `released` phase seems logical to me b/c all the project releases have been successfully created and images published. Up until now, we've been manually bumping the charts version after a release is created (when we thought of it 😄 ) - this PR mimics that so I don't see why it could be an issue.

**Comment by skitt** [comment]
> > Shouldn’t this take place during the installers phase, so that the tagged Helm charts have the updated contents?
> 
> I'm not sure what you mean by that. The `released` phase seems logical to me b/c all the project releases have been successfully created and images published. Up until now, we've been manually bumping the charts version after a release is created (when we thought of it smile ) - this PR mimics that so I don't see why it could be an issue.

It seems to me that the goal of the releases project, for each project, is to produce a tag which points to whatever counts as the release for that project. In the charts project, that should be the commit which has the correct state for the version being released. If we wait until the `released` phase, the tag has already been made, so the version update in the Makefile won’t be significant for that tag until the next release. Even the `installers` phase is actually too late for that since that’s when we decide on the commit that’s going to be tagged for the charts project.

I know that the current contents of the PR reflect what we do manually, but what we do manually isn’t great.

**Comment by tpantelis** [comment]
> It seems to me that the goal of the releases project, for each project, is to produce a tag which points to whatever counts as the release for that project. In the charts project, that should be the commit which has the correct state for the version being released. If we wait until the `released` phase, the tag has already been made, so the version update in the Makefile won’t be significant for that tag until the next release. Even the `installers` phase is actually too late for that since that’s when we decide on the commit that’s going to be tagged for the charts project.
> 
> I know that the current contents of the PR reflect what we do manually, but what we do manually isn’t great.

OK - I think I see your point. It seems we should create the **submariner-charts** PR along with the `subctl` PR in the `installers` phase.

**Review by skitt — APPROVED** [review]

File: scripts/do-release.sh [hunk 1]
```diff
@@ -272,6 +272,11 @@ function release_installers() {
     # Create a PR for subctl to use these versions
     local project="subctl"
     record_errors update_dependencies Subctl "${SUBCTL_CONSUMES[@]}"
+
+    # Create a PR in submariner-charts to update the CHARTS_VERSION in the Makefile
+    local project="submariner-charts"
+    local update_dependencies_extra=update_charts_versions
+    record_errors update_dependencies Charts
```
Comment (mkolesnik) [inline scripts/do-release.sh:279]: Shouldn't this be done in the same place we create the operator update PR?

Otherwise you're updating the version already after the tag is created.

Also, this will only update it on the branch you're releasing on, not on `devel`, which is fine I guess unless the intention was to update this on `devel`

Comment (tpantelis) [reply scripts/do-release.sh:279]: > Shouldn't this be done in the same place we create the operator update PR?

I'm not sure where the "best" place is. I supposed it could be in the `projects` phase too but **submariner-charts** is an installer so the `installers` phase seems appropriate along with **subctl**.

> Otherwise you're updating the version already after the tag is created.

I'm not sure what you mean here. Are you referring to the **submariner-charts** tag? Isn't that created in the `released` phase?

> Also, this will only update it on the branch you're releasing on, not on `devel`, which is fine I guess unless the intention was to update this on `devel`

The intention is update the release branch.

Comment (mkolesnik) [reply scripts/do-release.sh:279]: > > Shouldn't this be done in the same place we create the operator update PR?
> 
> I'm not sure where the "best" place is. I supposed it could be in the `projects` phase too but **submariner-charts** is an installer so the `installers` phase seems appropriate along with **subctl**.

Each stage is when those components get tagged, so on `installers` we tag `submariner-operator` and `submariner-charts` as those are out installers.
`subctl` is not an installer, it's a CLI, and as it relies on operator it has to be last, hence we create the relevant PR during the `installers` stage right after the operator is tagged and becomes available for `subctl` to use.

> 
> > Otherwise you're updating the version already after the tag is created.
> 
> I'm not sure what you mean here. Are you referring to the **submariner-charts** tag? Isn't that created in the `released` phase?

All these https://github.com/submariner-io/submariner-charts/tags are created on the `installers` stage, so since the chart "version" is actually the value of `CHARTS_VERSION`, we'd rather have it updated before the tag IIUC.

To put it differently - if I go and checkout tag `v0.15.1` in the project, which `CHARTS_VERSION` should I be seeing in the `Makefile`?

> 
> > Also, this will only update it on the branch you're releasing on, not on `devel`, which is fine I guess unless the intention was to update this on `devel`
> 
> The intention is update the release branch.

:+1:

Comment (tpantelis) [reply scripts/do-release.sh:279]: > subctl is not an installer, it's a CLI

It is a CLI installer, among other things. Otherwise, why did you name the phase "installers"? 😉 To me "installers" refers to components that install submariner in a k8s cluster which is what **submariner-charts** does , unless you intended a different meaning.

> if I go and checkout tag v0.15.1 in the project, which CHARTS_VERSION should I be seeing in the Makefile?

_0.15.1_.  I now see that `INSTALLER_PROJECTS ` includes **submariner-charts** and in `release_installers`:

`for_every_project create_project_release "${INSTALLER_PROJECTS[@]}"`

I assumed the **submariner-charts** tag was created in the `released` phase. In `release_released`:

```
for_every_project create_project_release "${PROJECTS[@]}"
```

and `PROJECTS=(admiral cloud-prepare lighthouse shipyard subctl submariner submariner-charts submariner-operator)`.  So it creates a project release for **submariner-charts** et al in more than one phase but I see that `create_project_release` checks if it already exists.

So.... it seems to me we could either move creating of the **submariner-charts** to `projects` or don't create the **submariner-charts** in the projects phase, ie remove it from `INSTALLER_PROJECTS`. If all else being equal, I'd rather do the latter since IMO semantically the `installers` phase fits better.
